#Exercise 1

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.1 MB 5.2 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
import numpy as np

In [4]:
coef = np.loadtxt('lab6_practice_coef.txt', delimiter=',')

In [5]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])

(17, 11)
Num rows: 17
Num cols: 11


In [6]:
model = ConcreteModel()

In [7]:
N = coef.shape[1]-1
M = coef.shape[0]-1

In [8]:
obj_coef = coef[0,:-1]
print(obj_coef.shape)
print(obj_coef)

(10,)
[0. 0. 0. 0. 0. 1. 1. 1. 1. 1.]


In [9]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)
print(constr_coef)

(16, 10)
[[ 85.  92.  45.  27.  31.   0.   0.   0.   0.   0.]
 [ 92.  54.  22.  20.   7.   0.   0.   0.   0.   0.]
 [ 96.  67.  29.  20.  11.   0.   0.   0.   0.   0.]
 [-91. -57. -33. -23. -12.   0.   0.   0.   0.   0.]
 [-99. -75. -26. -24. -41.   0.   0.   0.   0.   0.]
 [-98. -99. -57. -45. -65.   0.   0.   0.   0.   0.]
 [ -1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.  -1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.  -1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.  -1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.  -1.   0.   0.   0.   0.   1.]
 [  1.   0.   0.   0.   0.   1.   0.   0.   0.   0.]
 [  0.   1.   0.   0.   0.   0.   1.   0.   0.   0.]
 [  0.   0.   1.   0.   0.   0.   0.   1.   0.   0.]
 [  0.   0.   0.   1.   0.   0.   0.   0.   1.   0.]
 [  0.   0.   0.   0.   1.   0.   0.   0.   0.   1.]]


In [10]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)
print(constr_rhs)

(16,)
[1. 1. 1. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [11]:
col_indices_x = np.arange(int(N/2))
print(col_indices_x)
model.x = Var(col_indices_x)

[0 1 2 3 4]


In [12]:
col_indices_u = np.arange(int(N/2))
print(col_indices_u)
model.u = Var(col_indices_u, domain=NonNegativeReals)

[0 1 2 3 4]


In [13]:
row_indices = np.arange(M)
print(row_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]


In [14]:
model.constraints = ConstraintList()

In [15]:
for i in row_indices:
  model.constraints.add(summation(constr_coef[i][0:int(N/2)],model.x) + summation(constr_coef[i][int(N/2):N],model.u)  >= constr_rhs[i])

In [16]:
model.objective = Objective(expr = summation(obj_coef[0:int(N/2)],model.x) + summation(obj_coef[int(N/2):N],model.u), sense=minimize)

In [17]:
model.pprint()

3 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    u_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

2 Var Declarations
    u : Size=5, Index=u_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    x : Siz

In [18]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 148492 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [19]:
opt_cbc = SolverFactory('cbc')

In [20]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


##Ans 2 :

The Solver Status is Ok.

The Solver Termination Condition is optimal.

Yes, the solver yields an optimal solution.

In [21]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for j in col_indices_x:
    print('x[',j,']:', model.x[j].value)

for j in col_indices_u:
    print('u[',j,']:', model.u[j].value)

print('\nConstraints')
model.constraints.display()

Objective =  0.547955725

Decision Variables
x[ 0 ]: -0.046448575
x[ 1 ]: 0.18612441
x[ 2 ]: -0.17131802
x[ 3 ]: 0.0
x[ 4 ]: -0.14406472
u[ 0 ]: 0.046448575
u[ 1 ]: 0.18612441
u[ 2 ]: 0.17131802
u[ 3 ]: 0.0
u[ 4 ]: 0.14406472

Constraints
constraints : Size=16
    Key : Lower : Body               : Upper
      1 :   1.0 : 0.9999996250000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :  1.458337769999999 :  None
      4 :   1.0 : 1.0000002550000002 :  None
      5 :   1.0 :  1.000000215000001 :  None
      6 :   1.0 : 5.2549776999999995 :  None
      7 :   0.0 :         0.09289715 :  None
      8 :   0.0 :                0.0 :  None
      9 :   0.0 :         0.34263604 :  None
     10 :   0.0 :                0.0 :  None
     11 :   0.0 :         0.28812944 :  None
     12 :   0.0 :                0.0 :  None
     13 :   0.0 :         0.37224882 :  None
     14 :   0.0 :                0.0 :  None
     15 :   0.0 :                0.0 :  None
     16 :   0.0 :  

##Ans 3 :

Optimal Objective Function value : 0.547955725

$x_1: -0.046448575\\
x_2: 0.18612441\\
x_3: -0.17131802\\
x_4: 0.0\\
x_5: -0.14406472\\
u_1: 0.046448575\\
u_2: 0.18612441\\
u_3: 0.17131802\\
u_4: 0.0\\
u_5: 0.14406472$

Constraint Activities:

Constraint 1 is active by $\epsilon$ activity concept.

Constraint 2 is active by $\epsilon$ activity concept.

Constraint 3 is not active.

Constraint 4 is active by $\epsilon$ activity concept.

Constraint 5 is active by $\epsilon$ activity concept.

Constraint 6 is not active.

Constraint 7 is not active.

Constraint 8 is active by $\epsilon$ activity concept.

Constraint 9 is not active.

Constraint 10 is active by $\epsilon$ activity concept.

Constraint 11 is not active.

Constraint 12 is active by $\epsilon$ activity concept.

Constraint 13 is not active. 

Constraint 14 is active by $\epsilon$ activity concept.

Constraint 15 is active by $\epsilon$ activity concept.

Constraint 16 is active by $\epsilon$ activity concept.

***When we observe the values of $x_i$ with $u_i$, we see $u_i$ are the absolute value of $x_i$***

In [22]:
coef_OP2 = np.loadtxt('/content/lab6_practice_coef_OP2.txt',delimiter = ',')

In [23]:
print(coef_OP2.shape)
print('Num rows:',coef_OP2.shape[0])
print('Num cols:',coef_OP2.shape[1])

(7, 11)
Num rows: 7
Num cols: 11


In [24]:
model_OP2 = ConcreteModel()

In [25]:
n = coef_OP2.shape[1]-1
m = coef_OP2.shape[0]-1

In [26]:
obj_coef_OP2 = coef_OP2[0,:-1]
print(obj_coef_OP2)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


In [27]:
constr_coef_OP2 = coef_OP2[1:,:-1]
print(constr_coef_OP2)

[[ 85.  92.  45.  27.  31. -85. -92. -45. -27. -31.]
 [ 92.  54.  22.  20.   7. -92. -54. -22. -20.  -7.]
 [ 96.  67.  29.  20.  11. -96. -67. -29. -20. -11.]
 [-91. -57. -33. -23. -12.  91.  57.  33.  23.  12.]
 [-99. -75. -26. -24. -41.  99.  75.  26.  24.  41.]
 [-98. -99. -57. -45. -65.  98.  99.  57.  45.  65.]]


In [28]:
constr_rhs_OP2 = coef_OP2[1:,-1]
print(constr_rhs_OP2)

[1. 1. 1. 1. 1. 1.]


In [29]:
col_indices_a = np.arange(int(n/2))
model_OP2.a = Var(col_indices_a,domain = NonNegativeReals)

In [30]:
col_indices_b = np.arange(int(n/2))
model_OP2.b = Var(col_indices_b,domain = NonNegativeReals)

In [31]:
row_indices = np.arange(m)

In [32]:
model_OP2.constraints = ConstraintList()

In [33]:
for i in row_indices:
  model_OP2.constraints.add(summation(constr_coef_OP2[i][0:int(n/2)],model_OP2.a)+summation(constr_coef_OP2[i][int(n/2):N],model_OP2.b)>=constr_rhs_OP2[i])

In [34]:
model_OP2.objective = Objective(expr = summation(obj_coef_OP2[0:int(n/2)],model_OP2.a)+summation(obj_coef_OP2[int(n/2):n],model_OP2.b),sense = minimize)

In [35]:
model_OP2.pprint()

3 Set Declarations
    a_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    b_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}

2 Var Declarations
    a : Size=5, Index=a_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals
          4 :     0 :  None :  None : False :  True : NonNegativeReals
    b : Size=5, Index=b_index
        Key : Lowe

In [36]:
result = opt_cbc.solve(model_OP2)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


##Ans 9 :

The Solver Status is Ok.

The Solver Termination Condition is optimal.

Yes, the solver yields an optimal solution.

In [37]:
print('Objective = ', model_OP2.objective())

print('\nDecision Variables')
for j in col_indices_x:
    print('a[',j,']:', model_OP2.a[j].value)

for j in col_indices_u:
    print('b[',j,']:', model_OP2.b[j].value)

print('\nConstraints')
model.constraints.display()

Objective =  0.547955725

Decision Variables
a[ 0 ]: 0.0
a[ 1 ]: 0.18612441
a[ 2 ]: 0.0
a[ 3 ]: 0.0
a[ 4 ]: 0.0
b[ 0 ]: 0.046448575
b[ 1 ]: 0.0
b[ 2 ]: 0.17131802
b[ 3 ]: 0.0
b[ 4 ]: 0.14406472

Constraints
constraints : Size=16
    Key : Lower : Body               : Upper
      1 :   1.0 : 0.9999996250000001 :  None
      2 :   1.0 : 0.9999997599999997 :  None
      3 :   1.0 :  1.458337769999999 :  None
      4 :   1.0 : 1.0000002550000002 :  None
      5 :   1.0 :  1.000000215000001 :  None
      6 :   1.0 : 5.2549776999999995 :  None
      7 :   0.0 :         0.09289715 :  None
      8 :   0.0 :                0.0 :  None
      9 :   0.0 :         0.34263604 :  None
     10 :   0.0 :                0.0 :  None
     11 :   0.0 :         0.28812944 :  None
     12 :   0.0 :                0.0 :  None
     13 :   0.0 :         0.37224882 :  None
     14 :   0.0 :                0.0 :  None
     15 :   0.0 :                0.0 :  None
     16 :   0.0 :                0.0 :  None


##Ans 10 :

Optimal Objective Function value : 0.547955725

$a_1: 0.0\\
a_2: 0.18612441\\
a_3: 0.0\\
a_4: 0.0\\
a_5: 0.0\\
b_1: 0.046448575\\
b_2: 0.0\\
b_3: 0.17131802\\
b_4: 0.0\\
b_5: 0.14406472$

Constraint Activities:

Constraint 1 is active by $\epsilon$ activity concept.

Constraint 2 is active by $\epsilon$ activity concept.

Constraint 3 is not active.

Constraint 4 is active by $\epsilon$ activity concept.

Constraint 5 is active by $\epsilon$ activity concept.

Constraint 6 is not active.

Constraint 7 is not active.

Constraint 8 is active by $\epsilon$ activity concept.

Constraint 9 is not active.

Constraint 10 is active by $\epsilon$ activity concept.

Constraint 11 is not active.

Constraint 12 is active by $\epsilon$ activity concept.

Constraint 13 is not active. 

Constraint 14 is active by $\epsilon$ activity concept.

Constraint 15 is active by $\epsilon$ activity concept.

Constraint 16 is active by $\epsilon$ activity concept.



##Ans 11 :

Code is in the cell below:

In [39]:
for i in col_indices_x:
  print('x[',i+1,']:',model_OP2.a[i].value-model_OP2.b[i].value)

x[ 1 ]: -0.046448575
x[ 2 ]: 0.18612441
x[ 3 ]: -0.17131802
x[ 4 ]: 0.0
x[ 5 ]: -0.14406472


##Ans 12 :

The values of $x_i$ obtained from OP1 and OP2 are the same. We substitute $x_i$ with other variables to make our nonlinear program linear. But after solving, we obtain the values of x_i by putting the values accordingly.
Since the linear programs are equivalent, i.e. the solutions of both the programs are same, our original $x_i$ can be obtained.